### ISSUE TITLE
Correct way to use Steal on Node?

### ISSUE THREAD

**Issue by stucox** [stealjs/steal#800]
I have some code I wrote to use client-side, but would like to reuse it on the server. It currently depends on my Steal config — specifically a YAML `translate()` plugin. In general it’d be nice if I could have my Steal config apply both on client & server.

My naive attempt to run my code on Node with Steal in control was:

```js
// start-server.js
import steal from 'steal';
steal.startup({main: 'src/server'});
```
```sh
babel-node start-server.js
```

But my server-side code also includes modules which use Node builtins, which aren’t found.

I tried adding [steal-builtins](https://www.npmjs.com/package/steal-builtins) — but I realised that’s more designed to provide would-be-node modules in the browser — e.g. `require('http')` loads [http-browserify](https://github.com/substack/http-browserify), which realises it’s not running in a browser so throws a [‘`no window object present`’ error](https://github.com/substack/http-browserify/blob/master/index.js#L54).

I saw that #749 brings these into core — but looks like it’s still using http-browserify, so I’d guess the same issue would arise?

**Comment by matthewp** [comment]
The canonical way would be:

```js
var steal = require("steal");

steal.import("my-main").then( ...
```

steal.import will take care of making sure startup has been called.

As for your builtins issue, you might try using [node-locate](https://github.com/stealjs/node-locate) which *should* resolve to the Node's own builtins. This is an experimental module (intended for 1.1) so your results may vary, but it's worth a shot. Would be something like:

```js
var steal = require("steal");
var nodeLocate = require("@stealjs/node-locate");

// This configures steal to use the extension
nodeLocate(steal.System);

steal.import("my/main").then( ...
```

**Comment by stucox** [comment]
Thanks for the response.

Results:

1. I had `main: "src/client"` in my package.json — calling `steal.import("src/server")` in my server-side code still loads src/client.js, which isn’t what I want.

2. Removing `main` from package.json and explicitly calling `steal.import("src/client")` in my client-side code causes Steal to try and fetch an `index.js` file from the root of my repo — which didn’t exist. Added an empty index.js and now client-side just loads src/client.js and server-side just loads src/server.js, which is what I’m after (but feels dirty!)

3. Some paths are not being loaded correctly — e.g. I’m importing [snabbdom-to-html](https://github.com/acstll/snabbdom-to-html/tree/v2.1.3), for which Steal correctly loads its [src/index.js](https://github.com/acstll/snabbdom-to-html/blob/v2.1.3/src/index.js) entry point, but the `require('./init')` call within that file fails to find src/init.js within that package (it tries to load `init.js` from the root of my project). Is this likely to be a bug in node-locate? If the latter I can raise an issue in that repo and discuss further / do my best to contribute.

**Comment by matthewp** [comment]
Thanks, 1&2 are bugs, with `steal.import` we shouldn't first the `main` to load first. Could you file a new issue for that here? If so I can probably fix it soon.

3 I'm not sure about, can you recreate it in the browser?

**Comment by stucox** [comment]
Re. 1&2: Done: #802

3: I’ll try and create a minimal case to reproduce it.

**Comment by stucox** [comment]
Ok, here’s a test case — not with the same lib, but it looks like it might be the same cause:

https://gist.github.com/stucox/889767b4677818fbcefe35c1138a781e

**Comment by matthewp** [comment]
Thanks, will take a look.

**Comment by stucox** [comment]
Sorry about the silence — after a bit of absence, I tried the test case above with `0.16.38` and still the same result: Steal tries to request `localhost:8080/es5-ext/array/%23/clear.js`, when it should be `localhost:8080/node_modules/es5-ext/array/%23/clear.js`.

Have got as far as finding out it’s down to the path normalization — maybe it doesn’t deal with the `/%23/` (i.e. `/#/`) too well? (which is a really weird thing to be in a path anyway). Although I don’t think the other case I found these sorts of issues with had any such weird paths.

Will report back when I have more.
